daemon/api_notices.go

```diff
@@ -232,6 +246,104 @@ func allowedNoticeTypesForInterface(iface string) []state.NoticeType {
 	return types
 }
 
+func postNotices(c *Command, r *http.Request, user *auth.UserState) Response {
+	requestUID, err := uidFromRequest(r)
+	if err != nil {
+		return Forbidden("cannot determine UID of request, so cannot create notice")
+	}
+
+	var payload struct {
+		Action string `json:"action"`
+		Type   string `json:"type"`
+		Key    string `json:"key"`
+		// NOTE: Data and RepeatAfter fields are not needed for snap-run-inhibit notices.
+	}
+	decoder := json.NewDecoder(r.Body)
+	if err := decoder.Decode(&payload); err != nil {
+		return BadRequest("cannot decode request body: %v", err)
+	}
```

Comment:
This is definitely not a nice solution - this distracts from the purpose of the postNotices handler, and creates a lot of noise. Please move `var payload struct` out of the function, give it a more meaningful name and a docstring explaining why it's needed/purpose, and then create a function called `decode<>AndValidate`, where you move the validation logic to as well. Having a separate `decode` and `validate` function here is acceptable as well.

Reply:
this code is like this in pebble:

https://github.com/canonical/pebble/blob/master/internals/daemon/api_notices.go#L200

I have less of a strong opinion on this, as the struct is not used across many endpoints. Anyway if we change this we would need to backport the change to pebble

Reply:
this style is used in a few places in daemon fwiw

Reply:
I believe for snapd it makes more sense to have the `decode<>AndValidate` approach because we might have other notices (e.g. `custom` notices someday), but in I don't believe it make sense for pebble since the postNotices endpoint whole purpose is to support `custom` notices, it was built for specific purpose.

I understand the appeal of using a pattern that works for both snapd and pebble, but I think input from @benhoyt would be appreciated.

Reply:
Updated, thanks!